From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Matteo Collina <hello@matteocollina.com>
Date: Tue, 3 Jun 2025 11:09:40 +0200
Subject: [PATCH 36/74] http2: remove support for priority signaling

Signed-off-by: Matteo Collina <hello@matteocollina.com>
Co-authored-by: Antoine du Hamel <duhamelantoine1995@gmail.com>
Refs: https://datatracker.ietf.org/doc/html/rfc9113#section-5.3.1
PR-URL: https://github.com/nodejs/node/pull/58293
Backport-PR-URL: https://github.com/nodejs/node/pull/63195
Reviewed-By: Marco Ippolito <marcoippolito54@gmail.com>
Reviewed-By: Tim Perry <pimterry@gmail.com>
Reviewed-By: Rafael Gonzaga <rafael.nunu@hotmail.com>
Reviewed-By: Antoine du Hamel <duhamelantoine1995@gmail.com>
(cherry picked from commit ea5dc6b529b274b8835ea76951535d07d2349495)
(cherry picked from commit 14e0c838b08ff0f293beeeb1d3185175cee39d64)
Backport-Notes: v18.x never had the DEP0194 documentation-only
  deprecation, so DEP0194 is introduced here directly as End-of-Life. This
  is required to move to nghttp2 1.69.0, which drops RFC 7540 priorities.
---
 doc/api/deprecations.md                       | 14 +++++
 doc/api/http2.md                              | 44 ++++++-------
 lib/internal/http2/core.js                    | 61 ++++++-------------
 lib/internal/util.js                          | 12 ++++
 ...st-http2-client-priority-before-connect.js |  5 ++
 ...est-http2-client-request-options-errors.js |  1 -
 .../test-http2-client-set-priority.js         |  8 ++-
 test/parallel/test-http2-priority-cycle-.js   |  5 ++
 test/parallel/test-http2-priority-event.js    | 19 +++---
 ...est-http2-server-stream-session-destroy.js |  5 ++
 10 files changed, 97 insertions(+), 77 deletions(-)

diff --git a/doc/api/deprecations.md b/doc/api/deprecations.md
index b08160b748..b605abeafe 100644
--- a/doc/api/deprecations.md
+++ b/doc/api/deprecations.md
@@ -3347,9 +3347,23 @@ Type: Documentation-only
 The [`dirent.path`][] is deprecated due to its lack of consistency across
 release lines. Please use [`dirent.parentPath`][] instead.
 
+### DEP0194: HTTP/2 priority signaling
+
+<!-- YAML
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/58293
+    description: End-of-Life.
+-->
+
+Type: End-of-Life
+
+The support for priority signaling has been removed following its deprecation in the [RFC 9113][].
+
 [NIST SP 800-38D]: https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf
 [RFC 6066]: https://tools.ietf.org/html/rfc6066#section-3
 [RFC 8247 Section 2.4]: https://www.rfc-editor.org/rfc/rfc8247#section-2.4
+[RFC 9113]: https://datatracker.ietf.org/doc/html/rfc9113#section-5.3.1
 [WHATWG URL API]: url.md#the-whatwg-url-api
 [`"exports"` or `"main"` entry]: packages.md#main-entry-point-export
 [`'uncaughtException'`]: process.md#event-uncaughtexception
diff --git a/doc/api/http2.md b/doc/api/http2.md
index bc63db0dcc..9a58c834de 100644
--- a/doc/api/http2.md
+++ b/doc/api/http2.md
@@ -929,6 +929,11 @@ The `'origin'` event is only emitted when using a secure TLS connection.
 
 <!-- YAML
 added: v8.4.0
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/58293
+    description: The `weight` option is now ignored, setting it will trigger a
+                 runtime warning.
 -->
 
 * `headers` {HTTP/2 Headers Object}
@@ -943,9 +948,6 @@ added: v8.4.0
     **Default:** `false`.
   * `parent` {number} Specifies the numeric identifier of a stream the newly
     created stream is dependent on.
-  * `weight` {number} Specifies the relative dependency of a stream in relation
-    to other streams with the same `parent`. The value is a number between `1`
-    and `256` (inclusive).
   * `waitForTrailers` {boolean} When `true`, the `Http2Stream` will emit the
     `'wantTrailers'` event after the final `DATA` frame has been sent.
   * `signal` {AbortSignal} An AbortSignal that may be used to abort an ongoing
@@ -1299,22 +1301,18 @@ numeric stream identifier.
 
 <!-- YAML
 added: v8.4.0
+deprecated: REPLACEME
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/58293
+    description: This method no longer sets the priority of the stream. Using it
+                 now triggers a runtime warning.
 -->
 
-* `options` {Object}
-  * `exclusive` {boolean} When `true` and `parent` identifies a parent Stream,
-    this stream is made the sole direct dependency of the parent, with
-    all other existing dependents made a dependent of this stream. **Default:**
-    `false`.
-  * `parent` {number} Specifies the numeric identifier of a stream this stream
-    is dependent on.
-  * `weight` {number} Specifies the relative dependency of a stream in relation
-    to other streams with the same `parent`. The value is a number between `1`
-    and `256` (inclusive).
-  * `silent` {boolean} When `true`, changes the priority locally without
-    sending a `PRIORITY` frame to the connected peer.
-
-Updates the priority for this `Http2Stream` instance.
+> Stability: 0 - Deprecated: support for priority signaling has been deprecated
+> in the [RFC 9113][] and is no longer supported in Node.js.
+
+Empty method, only there to maintain some backward compatibility.
 
 #### `http2stream.rstCode`
 
@@ -1400,6 +1398,11 @@ req.setTimeout(5000, () => req.close(NGHTTP2_CANCEL));
 
 <!-- YAML
 added: v8.4.0
+changes:
+  - version: REPLACEME
+    pr-url: https://github.com/nodejs/node/pull/58293
+    description: The `state.weight` property is now always set to 16 and
+                 `sumDependencyWeight` is always set to 0.
 -->
 
 Provides miscellaneous information about the current state of the
@@ -1413,10 +1416,8 @@ Provides miscellaneous information about the current state of the
   * `localClose` {number} `1` if this `Http2Stream` has been closed locally.
   * `remoteClose` {number} `1` if this `Http2Stream` has been closed
     remotely.
-  * `sumDependencyWeight` {number} The sum weight of all `Http2Stream`
-    instances that depend on this `Http2Stream` as specified using
-    `PRIORITY` frames.
-  * `weight` {number} The priority weight of this `Http2Stream`.
+  * `sumDependencyWeight` {number} Legacy property, always set to `0`.
+  * `weight` {number} Legacy property, always set to `16`.
 
 A current state of this `Http2Stream`.
 
@@ -4177,6 +4178,7 @@ you need to implement any fall-back behavior yourself.
 [RFC 7838]: https://tools.ietf.org/html/rfc7838
 [RFC 8336]: https://tools.ietf.org/html/rfc8336
 [RFC 8441]: https://tools.ietf.org/html/rfc8441
+[RFC 9113]: https://datatracker.ietf.org/doc/html/rfc9113#section-5.3.1
 [Sensitive headers]: #sensitive-headers
 [`'checkContinue'`]: #event-checkcontinue
 [`'connect'`]: #event-connect
diff --git a/lib/internal/http2/core.js b/lib/internal/http2/core.js
index 9ba39d49dd..e5af2ecbe6 100644
--- a/lib/internal/http2/core.js
+++ b/lib/internal/http2/core.js
@@ -38,6 +38,8 @@ const {
   customInspectSymbol: kInspect,
   kEmptyObject,
   promisify,
+  deprecate,
+  deprecateProperty,
 } = require('internal/util');
 
 assertCrypto();
@@ -706,6 +708,11 @@ function onGoawayData(code, lastStreamID, buf) {
   }
 }
 
+// TODO(aduh95): remove this in future semver-major
+const deprecateWeight = deprecateProperty('weight',
+                                          'Priority signaling has been deprecated as of RFC 1993.',
+                                          'DEP0194');
+
 // When a ClientHttp2Session is first created, the socket may not yet be
 // connected. If request() is called during this time, the actual request
 // will be deferred until the socket is ready to go.
@@ -734,12 +741,14 @@ function requestOnConnect(headers, options) {
   if (options.waitForTrailers)
     streamOptions |= STREAM_OPTION_GET_TRAILERS;
 
+  deprecateWeight(options);
+
   // `ret` will be either the reserved stream ID (if positive)
   // or an error code (if negative)
   const ret = session[kHandle].request(headers,
                                        streamOptions,
                                        options.parent | 0,
-                                       options.weight | 0,
+                                       NGHTTP2_DEFAULT_WEIGHT,
                                        !!options.exclusive);
 
   // In an error condition, one of three possible response codes will be
@@ -778,11 +787,7 @@ function requestOnConnect(headers, options) {
 //
 // Also sets the default priority options if they are not set.
 const setAndValidatePriorityOptions = hideStackFrames((options) => {
-  if (options.weight === undefined) {
-    options.weight = NGHTTP2_DEFAULT_WEIGHT;
-  } else if (typeof options.weight !== 'number') {
-    throw new ERR_INVALID_ARG_VALUE('options.weight', options.weight);
-  }
+  deprecateWeight(options);
 
   if (options.parent === undefined) {
     options.parent = 0;
@@ -839,25 +844,6 @@ function submitSettings(settings, callback) {
   }
 }
 
-// Submits a PRIORITY frame to be sent to the remote peer
-// Note: If the silent option is true, the change will be made
-// locally with no PRIORITY frame sent.
-function submitPriority(options) {
-  if (this.destroyed)
-    return;
-  this[kUpdateTimer]();
-
-  // If the parent is the id, do nothing because a
-  // stream cannot be made to depend on itself.
-  if (options.parent === this[kID])
-    return;
-
-  this[kHandle].priority(options.parent | 0,
-                         options.weight | 0,
-                         !!options.exclusive,
-                         !!options.silent);
-}
-
 // Submit a GOAWAY frame to be sent to the remote peer.
 // If the lastStreamID is set to <= 0, then the lastProcStreamID will
 // be used. The opaqueData must either be a typed array or undefined
@@ -2237,25 +2223,6 @@ class Http2Stream extends Duplex {
     }
   }
 
-  priority(options) {
-    if (this.destroyed)
-      throw new ERR_HTTP2_INVALID_STREAM();
-
-    assertIsObject(options, 'options');
-    options = { ...options };
-    setAndValidatePriorityOptions(options);
-
-    const priorityFn = FunctionPrototypeBind(submitPriority, this, options);
-
-    // If the handle has not yet been assigned, queue up the priority
-    // frame to be sent as soon as the ready event is emitted.
-    if (this.pending) {
-      this.once('ready', priorityFn);
-      return;
-    }
-    priorityFn();
-  }
-
   sendTrailers(headers) {
     if (this.destroyed || this.closed)
       throw new ERR_HTTP2_INVALID_STREAM();
@@ -2410,6 +2377,12 @@ class Http2Stream extends Duplex {
   }
 }
 
+// TODO(aduh95): remove this in future semver-major
+Http2Stream.prototype.priority = deprecate(function priority(options) {
+  if (this.destroyed)
+    throw new ERR_HTTP2_INVALID_STREAM();
+}, 'http2Stream.priority is longer supported after priority signalling was deprecated in RFC 1993', 'DEP0194');
+
 function callTimeout(self, kSession) {
   // If the session is destroyed, this should never actually be invoked,
   // but just in case...
diff --git a/lib/internal/util.js b/lib/internal/util.js
index aedfa2e988..2d7cff4293 100644
--- a/lib/internal/util.js
+++ b/lib/internal/util.js
@@ -135,6 +135,17 @@ function isPendingDeprecation() {
     !getOptionValue('--no-deprecation');
 }
 
+function deprecateProperty(key, msg, code, isPendingDeprecation) {
+  const emitDeprecationWarning = getDeprecationWarningEmitter(
+    code, msg, undefined, false, isPendingDeprecation,
+  );
+  return (options) => {
+    if (key in options) {
+      emitDeprecationWarning();
+    }
+  };
+}
+
 // Internal deprecator for pending --pending-deprecation. This can be invoked
 // at snapshot building time as the warning permission is only queried at
 // run time.
@@ -808,6 +819,7 @@ module.exports = {
   defineLazyProperties,
   defineReplaceableLazyAttribute,
   deprecate,
+  deprecateProperty,
   emitExperimentalWarning,
   exposeInterface,
   exposeLazyInterfaces,
diff --git a/test/parallel/test-http2-client-priority-before-connect.js b/test/parallel/test-http2-client-priority-before-connect.js
index 7aa13a5e45..2620126585 100644
--- a/test/parallel/test-http2-client-priority-before-connect.js
+++ b/test/parallel/test-http2-client-priority-before-connect.js
@@ -5,6 +5,11 @@ if (!common.hasCrypto)
   common.skip('missing crypto');
 const h2 = require('http2');
 
+common.expectWarning(
+  'DeprecationWarning',
+  'http2Stream.priority is longer supported after priority signalling was deprecated in RFC 1993',
+  'DEP0194');
+
 const server = h2.createServer();
 
 // We use the lower-level API here
diff --git a/test/parallel/test-http2-client-request-options-errors.js b/test/parallel/test-http2-client-request-options-errors.js
index f2950fe4d2..e24619a3a5 100644
--- a/test/parallel/test-http2-client-request-options-errors.js
+++ b/test/parallel/test-http2-client-request-options-errors.js
@@ -12,7 +12,6 @@ const { inspect } = require('util');
 
 const optionsToTest = {
   endStream: 'boolean',
-  weight: 'number',
   parent: 'number',
   exclusive: 'boolean',
   silent: 'boolean'
diff --git a/test/parallel/test-http2-client-set-priority.js b/test/parallel/test-http2-client-set-priority.js
index c41ec99031..7900c1173f 100644
--- a/test/parallel/test-http2-client-set-priority.js
+++ b/test/parallel/test-http2-client-set-priority.js
@@ -6,10 +6,16 @@ if (!common.hasCrypto)
 const assert = require('assert');
 const http2 = require('http2');
 
+common.expectWarning(
+  'DeprecationWarning',
+  'Priority signaling has been deprecated as of RFC 1993.',
+  'DEP0194');
+
 const checkWeight = (actual, expect) => {
   const server = http2.createServer();
   server.on('stream', common.mustCall((stream, headers, flags) => {
-    assert.strictEqual(stream.state.weight, expect);
+    assert.strictEqual(stream.state.sumDependencyWeight, 0);
+    assert.strictEqual(stream.state.weight, 16);
     stream.respond();
     stream.end('test');
   }));
diff --git a/test/parallel/test-http2-priority-cycle-.js b/test/parallel/test-http2-priority-cycle-.js
index af0d66d834..49a6de1029 100644
--- a/test/parallel/test-http2-priority-cycle-.js
+++ b/test/parallel/test-http2-priority-cycle-.js
@@ -7,6 +7,11 @@ const assert = require('assert');
 const http2 = require('http2');
 const Countdown = require('../common/countdown');
 
+common.expectWarning(
+  'DeprecationWarning',
+  'http2Stream.priority is longer supported after priority signalling was deprecated in RFC 1993',
+  'DEP0194');
+
 const server = http2.createServer();
 const largeBuffer = Buffer.alloc(1e4);
 
diff --git a/test/parallel/test-http2-priority-event.js b/test/parallel/test-http2-priority-event.js
index 41ec6534b2..dad9433620 100644
--- a/test/parallel/test-http2-priority-event.js
+++ b/test/parallel/test-http2-priority-event.js
@@ -3,21 +3,18 @@
 const common = require('../common');
 if (!common.hasCrypto)
   common.skip('missing crypto');
-const assert = require('assert');
 const h2 = require('http2');
 
+common.expectWarning(
+  'DeprecationWarning',
+  'http2Stream.priority is longer supported after priority signalling was deprecated in RFC 1993',
+  'DEP0194');
+
 const server = h2.createServer();
 
 // We use the lower-level API here
 server.on('stream', common.mustCall(onStream));
 
-function onPriority(stream, parent, weight, exclusive) {
-  assert.strictEqual(stream, 1);
-  assert.strictEqual(parent, 0);
-  assert.strictEqual(weight, 1);
-  assert.strictEqual(exclusive, false);
-}
-
 function onStream(stream, headers, flags) {
   stream.priority({
     parent: 0,
@@ -33,7 +30,7 @@ function onStream(stream, headers, flags) {
 
 server.listen(0);
 
-server.on('priority', common.mustCall(onPriority));
+server.on('priority', common.mustNotCall());
 
 server.on('listening', common.mustCall(() => {
 
@@ -48,7 +45,9 @@ server.on('listening', common.mustCall(() => {
     });
   });
 
-  req.on('priority', common.mustCall(onPriority));
+  // The priority event is not supported anymore by nghttp2
+  // since 1.65.0.
+  req.on('priority', common.mustNotCall());
 
   req.on('response', common.mustCall());
   req.resume();
diff --git a/test/parallel/test-http2-server-stream-session-destroy.js b/test/parallel/test-http2-server-stream-session-destroy.js
index 34b22fdfbd..88fcb3b3cd 100644
--- a/test/parallel/test-http2-server-stream-session-destroy.js
+++ b/test/parallel/test-http2-server-stream-session-destroy.js
@@ -6,6 +6,11 @@ if (!common.hasCrypto)
 const assert = require('assert');
 const h2 = require('http2');
 
+common.expectWarning(
+  'DeprecationWarning',
+  'http2Stream.priority is longer supported after priority signalling was deprecated in RFC 1993',
+  'DEP0194');
+
 const server = h2.createServer();
 
 server.on('stream', common.mustCall((stream) => {
